src/borg/_hashindex.c

```diff
@@ -196,6 +196,7 @@ hashindex_lookup(HashIndex *index, const unsigned char *key, int *start_idx)
          * upper_limit and min_empty.
          */
         assert(idx != start);
+        if(idx == start) *(char *)NULL;  // BOOM!
```

Comment:
hmm, does the compiler optimize that away?

Reply:
also tried: `*(char *)NULL = 0;` - nothing happens, just hangs.

Reply:
but still the `assert(false);` does not abort the process.

Reply:
But the assertion doesn't cause an error?  That's annoying.

In any case, I didn't expect _lookup or _set to be called on a compacted hash_table.  I guess we'll need to put back the test for idx == start, and break out of the loop if that happens.

Reply:
Yeah, i do return with -2 in that case now.

assert() is documented to log a warning to stderr and call abort() to terminate the process.
the first i maybe don't see due to pytest, but the latter does not happen.

Reply:
BTW, i used adding a new element to the compact hashtable only as a way to trigger a resize/rebuild (that API is not available from Python otherwise).

Reply:
Are you building with -DNDEBUG? That would remove assertions.

Reply:
@horazont I searched for NDEBUG in the project, but couldn't find anything.

Reply:
See #7329.